### DISCUSSION TITLE
Having trouble receiving UI events

### DISCUSSION THREAD

**Discussion by smolck** [esensar/neovim-java#118] [Q&A] [answered]
Hello! So, I'd like to make a Neovim GUI which uses this library, but I'm having trouble getting things working. The code in question is essentially this (written in Kotlin since I prefer it to Java; I've also left out the imports to focus on the code, but let me know if you want those as well):

```kotlin
class UiEventSubscriber : Flow.Subscriber<NeovimRedrawEvent> {
    override fun onNext(item: NeovimRedrawEvent?) {
        println("Notif name: ${item?.notificationName}")
        TODO("Not yet implemented")
    }

    override fun onSubscribe(subscription: Flow.Subscription?) {
        TODO("Not yet implemented")
    }

    override fun onError(throwable: Throwable?) {
        TODO("Not yet implemented")
    }

    override fun onComplete() {
        TODO("Not yet implemented")
    }

}

fun main() {
    val process = ProcessBuilder("nvim", "--embed").start()
    val connection = ProcessRpcConnection(process)

    val rpcStreamer = ReactiveRpcClient.createDefaultInstance()
    rpcStreamer.attach(connection)

    val nvim = NeovimStreamApi(rpcStreamer)
    val notifHandler = NeovimStreamNotificationHandler(rpcStreamer)

    val subscriber = UiEventSubscriber()
    notifHandler.uiEvents().subscribe(subscriber)
    nvim.attachUi(80, 80, UiOptions.FULL_UI)

    nvim.executeCommand("e ~/path/to/file/on/system").thenAccept {
        nvim.currentBuffer.thenAccept {
            it.getLines(0, -1, false).thenAccept { lines ->
                println("${lines.joinToString()}")
            }
        }
    }
}
```

There are two problems I've run into with the above code. First, the `getLines` call doesn't seem to work; the lines from the buffer are never printed out. This seems to have something to do with `rpcStreamer` being involved, because if I get the nvim connection with `val nvim = NeovimApis.getApiForConnection(connection)` the command works fine and the opened buffer's contents are printed as expected.

Second, there appears to be an issue with `notifHandler.uiEvents().subscribe(subscriber)`. I keep getting this strange error from using it: 
```
Caused by: com.fasterxml.jackson.databind.JsonMappingException: 'gen' is expected to be MessagePackGenerator but it's class com.fasterxml.jackson.databind.util.TokenBuffer (through reference chain: java.util.ArrayList[23]->java.util.ArrayList[1]->java.util.ArrayList[1])`
```
Here's the full stacktrace for the example above (in a gist so as not to make this comment overly long): https://gist.github.com/smolck/e466adee2fd8b91d549468ecb37e8a53

Hopefully these issues are straightforward to fix, because until I know how to fix them I'm not sure I can get to working on this Neovim GUI I'd like to build. Thanks in advance!

**Comment by esensar** [comment, marked as answer]
Hello! Thank you for reporting this issue. I hope library is usable for you.

I have checked out the code you posted in the question and it seems that all issues come from notification handler. I believe there is a bug in library (the stacktrace you posted) which I will try to resolve quickly.

Regarding the `rpcStreamer` vs `NeovimApis`, it is my bad for bad API design. In the meantime you can use a workaround, using code similar to `NeovimApis`:
```
var rpcClient = new RpcClient.Builder()
        .withObjectMapper(NeovimJacksonModule.createNeovimObjectMapper()).build();
var reactiveRpcStreamer = ReactiveRpcClient.createDefaultInstanceWithCustomStreamer(rpcClient);
```

Key difference here is usage of `NeovimJacksonModule.createNeovimObjectMapper` which allows usage of `Buffer`, `Window` and `Tabpage` types with API.

I will need to look into issues with notification handlers. Also in the meantime, for future debugging, adding logger implementation to see library logs could be useful to you:
```
    implementation("org.slf4j:slf4j-simple:1.7.25")
```

To enable debug logs (which show all responses received from nvim instance), create `simplelogger.properties` file in your `resources` dir, with following contents:
```
org.slf4j.simpleLogger.defaultLogLevel=debug
```

I would also recommend using:
```
val process = ProcessBuilder("nvim", "--embed", "-u", "NONE").start()
```
Until your GUI has means of handling inputs, since in my case, loading my `vimrc` caused `nvim` to block (ignore all requests) until an input that was requested was complete.

**Reply by smolck** [reply]
Coupled with the new version and the change from your answer below, it works! Thank you very much for your help!

**Reply by esensar** [reply]
Not sure where you fetched latest version from. I had some minor issues deploying to Maven Central, but it should be resolved soon.

**Reply by esensar** [reply]
Also, thank you very much for the detailed question. While investigating this I found some more issues that need to be fixed in the library and also realized how some parts of the library are barely usable :smile:

**Reply by esensar** [reply]
Since 0.2.3, code block from above no longer works:
```
var rpcClient = new RpcClient.Builder()
        .withObjectMapper(NeovimJacksonModule.createNeovimObjectMapper()).build();
var reactiveRpcStreamer = ReactiveRpcClient.createDefaultInstanceWithCustomStreamer(rpcClient);
```

It can be easily resolved by replacing `NeovimJacksonModule.createNeovimObjectMapper()` with `ObjectMappers.defaultNeovimMapper()`, but it is recommended to instead use new utilities provided by `NeovimApis`:
```
NeovimApis.getNeovimReactiveRpcStreamer() // For reactive RPC client with neovim types support
NeovimApis.getNeovimRpcStream() // For regular RPC client with neovim types support
```

In the future it is planned to remove the need for any of these (these will remain available if necessary), by moving Neovim extensions types support to core modules, therefore removing need for customization of `RpcClient` to make it work with Neovim extensions types.

**Comment by esensar** [comment]
I have released version `0.2.1` which should fix issues with `UIEvent`. Your code should work after updating to `0.2.1` with one change:
```
override fun onSubscribe(subscription: Flow.Subscription?) {
        subscription?.request(Long.MAX_VALUE)
}
```
Otherwise no events will be received by this subscriber. You can of course request some other value, this requests all future events.

**Reply by smolck** [reply]
Thank you, that works! Also, might I ask what exactly `subscription?.request(Long.MAX_VALUE)` is doing/what is the purpose of it?

**Reply by esensar** [reply]
Honestly, I just took that as default when it comes to Java 9 Flow.
Since Java 9 Flows are pull based (meaning you only get events that you explicitly request), you need to request number of items. Doing it this way pretty much tells it to take all events. You could request just one, and after processing that one, ask another one with something like:
```
// Store subscription
lateinit var subscription: Flow.Subscription?

override fun onSubscribe(subscription: Flow.Subscription?) {
    this.subscription = subscription
    subscription?.request(1)
}

override fun onNext(item: T?) {
    // process item
    // ...
    // request more
    this.subscription?.request(1)
}
```
